internal/controllers/core/clustercatalog_controller.go

```diff
@@ -72,25 +73,38 @@ func (r *ClusterCatalogReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	}
 	reconciledCatsrc := existingCatsrc.DeepCopy()
 	res, reconcileErr := r.reconcile(ctx, reconciledCatsrc)
+	updateError := reconcileErr
 
-	// Update the status subresource before updating the main object. This is
-	// necessary because, in many cases, the main object update will remove the
-	// finalizer, which will cause the core Kubernetes deletion logic to
-	// complete. Therefore, we need to make the status update prior to the main
-	// object update to ensure that the status update can be processed before
-	// a potential deletion.
-	if !equality.Semantic.DeepEqual(existingCatsrc.Status, reconciledCatsrc.Status) {
-		if updateErr := r.Client.Status().Update(ctx, reconciledCatsrc); updateErr != nil {
-			return res, apimacherrors.NewAggregate([]error{reconcileErr, updateErr})
+	// Do checks before any Update()s, as Update() may modify the resource structure!
+	updateStatus := !equality.Semantic.DeepEqual(existingCatsrc.Status, reconciledCatsrc.Status)
+	updateFinalizers := !equality.Semantic.DeepEqual(existingCatsrc.Finalizers, reconciledCatsrc.Finalizers)
+	unexpectedFieldsChanged := checkForUnexpectedFieldChange(*&existingCatsrc, *reconciledCatsrc)
+
+	if unexpectedFieldsChanged {
+		panic("spec or metadata changed by reconciler")
+	}
+
+	// Save the finalizers off to the side. If we update the status, the reconciledExt will be updated
+	// to contain the new state of the ClusterExtension, which contains the status update, but (critically)
+	// does not contain the finalizers. After the status update, we need to re-add the finalizers to the
+	// reconciledExt before updating the object.
```

Comment:
s/reconciledExt/reconciledCatsrc/g:
```suggestion
	// Save the finalizers off to the side. If we update the status, the reconciledCatsrc will be updated
	// to contain the new state of the ClusterCatalog, which contains the status update, but (critically)
	// does not contain the finalizers. After the status update, we need to re-add the finalizers to the
	// reconciledCatsrc before updating the object.
```